cabal-install/Distribution/Client/Run.hs

```diff
@@ -42,23 +45,60 @@ import System.FilePath                       ((<.>), (</>))
 -- forwarded to it.
 splitRunArgs :: LocalBuildInfo -> [String] -> IO (Executable, [String])
 splitRunArgs lbi args =
-  case exes of
-    []    -> die "Couldn't find any executables."
-    [exe] -> case args of
-      []                        -> return (exe, [])
-      (x:xs) | x == exeName exe -> return (exe, xs)
-             | otherwise        -> return (exe, args)
-    _     -> case args of
-      []     -> die $ "This package contains multiple executables. "
-                ++ "You must pass the executable name as the first argument "
-                ++ "to 'cabal run'."
-      (x:xs) -> case find (\exe -> exeName exe == x) exes of
-        Nothing  -> die $ "No executable named '" ++ x ++ "'."
-        Just exe -> return (exe, xs)
+  case whichExecutable of -- Either err (wasManuallyChosen, exe, restParams)
+    Left err               -> do printMaybeWarningWithAdditional ""
+                                 die err
+    Right (True, exe, xs)  -> return (exe, xs)
+    Right (False, exe, xs) -> do printMaybeWarningWithAdditional
+                                   $ " Interpreting all parameters to `run`"
+                                  ++ " as a parameter to the"
+                                  ++ " default executable."
+                                 return (exe, xs)
   where
     pkg_descr = localPkgDescr lbi
-    exes      = executables pkg_descr
-
+    whichExecutable :: Either String  -- error string
+                              ( Bool       -- if it was manually chosen
+                              , Executable -- the executable
+                              , [String]   -- the remaining parameters
+                              )
+    whichExecutable = case (enabledExes, args) of
+      ([]   , _)           -> Left "Couldn't find any enabled executables."
+      ([exe], [])          -> return (False, exe, [])
+      ([exe], (x:xs))
+        | x == exeName exe -> return (True, exe, xs)
+        | otherwise        -> return (False, exe, args)
+      (_    , [])          -> Left
+        $ "This package contains multiple executables. "
+        ++ "You must pass the executable name as the first argument "
+        ++ "to 'cabal run'."
+      (_    , (x:xs))      ->
+        case find (\exe -> exeName exe == x) enabledExes of
+          Nothing  -> Left $ "No executable named '" ++ x ++ "'."
+          Just exe -> return (True, exe, xs)
+      where
+        enabledExes = filter (buildable . buildInfo) (executables pkg_descr)
+    printMaybeWarningWithAdditional :: String -> IO ()
```

Comment:
I think it's simpler to just make this `Maybe String` and then use e.g. `notice $ concat (maybeToList mbWarn ++ myWarn)`.